File: src/DawnlightSearch/Ui_change_advanced_setting_dialog.py
from __future__ import absolute_import

from DawnlightSearch._Global_Qt_import import *

# import app_recource_rc

# from Ui_Ui_advanced_setting_dialog import Ui_Dialog as EditSettingDialog_base_class
import os
EditSettingDialog_base_class, _ = uic.loadUiType(
    os.path.join(os.path.dirname(os.path.abspath(__file__)), "Ui_advanced_setting_dialog.ui"))
# EditSettingDialog_base_class, _ = uic.loadUiType("Ui_advanced_setting_dialog.ui")
class EditSettingDialog(QDialog, EditSettingDialog_base_class):

    def __init__(self, ORGANIZATION_NAME, ALLICATION_NAME, DATABASE_FILE_NAME, TEMP_DB_NAME, parent=None):
        self.ORGANIZATION_NAME = ORGANIZATION_NAME
        self.ALLICATION_NAME = ALLICATION_NAME

        super(self.__class__, self).__init__(parent)
        self.setupUi(self)

        settings = QSettings(QSettings.IniFormat, QSettings.UserScope, self.ORGANIZATION_NAME, self.ALLICATION_NAME)
        print settings.fileName()
        # print settings.value('Query_Chunk_Size', type=int, defaultValue=10000)

        self.spinBox_label_query_chunk_size.setValue(
            settings.value('Query_Chunk_Size', type=int, defaultValue=10000)
        )
        self.spinBox_model_max_items.setValue(
            settings.value('Max_Items_in_List', type=int, defaultValue=3000)
        )
        self.spinBox_query_limit.setValue(
            settings.value('Query_limit', type=int, defaultValue=100)
        )
        self.lineEdit_Database_File_Name.setText(DATABASE_FILE_NAME)
        self.lineEdit_Temp_Database_File_Name.setText(TEMP_DB_NAME)
        self.lineEdit_Database_File_Name.textEdited.connect(self.on_lineEdit_Database_File_Name_edited)
        self.lineEdit_Temp_Database_File_Name.textEdited.connect(self.on_Temp_lineEdit_Database_File_Name_edited)

        self.toolButton_Database_File_Name.released.connect(self.on_change_Database_File)
        self.toolButton_Temp_Database_File_Name.released.connect(self.on_change_Temp_Database_File)

    @pyqtSlot(str)
    def on_lineEdit_Database_File_Name_edited(self,path):
        if os.path.exists(os.path.dirname(path)):
            self.lineEdit_Database_File_Name.setStyleSheet(
                "");
        else:
            self.lineEdit_Database_File_Name.setStyleSheet(
                "QLineEdit { color: red;  }")

    @pyqtSlot(str)
    def on_Temp_lineEdit_Database_File_Name_edited(self, path):
        if os.path.exists(os.path.dirname(path)):
            self.lineEdit_Temp_Database_File_Name.setStyleSheet(
                "");
        else:
            self.lineEdit_Temp_Database_File_Name.setStyleSheet(
                "QLineEdit { color: red;  }");
    @pyqtSlot()
    def on_change_Database_File(self):
        folder = str(QFileDialog.getExistingDirectory(self, "Select Directory of Database File",
                                                      os.path.dirname(self.lineEdit_Database_File_Name.text())))
        if folder:
            filename = os.path.basename(self.lineEdit_Database_File_Name.text())
            self.lineEdit_Database_File_Name.setText(os.path.join(folder, filename))

    @pyqtSlot()
    def on_change_Temp_Database_File(self):
        folder = str(QFileDialog.getExistingDirectory(self, "Select Directory of Temp Database File",
                                                      os.path.dirname(self.lineEdit_Temp_Database_File_Name.text())))
        if folder:
            filename = os.path.basename(self.lineEdit_Temp_Database_File_Name.text())
            self.lineEdit_Temp_Database_File_Name.setText(os.path.join(folder, filename))

    def _getSettings(self):
        pass
        self.spinBox_label_query_chunk_size.value()
        return self.spinBox_label_query_chunk_size.value()

    # static method to create the dialog and return (date, time, accepted)
    @staticmethod
    def getSetting(ORGANIZATION_NAME, ALLICATION_NAME, DATABASE_FILE_NAME, TEMP_DB_NAME, parent=None):
        dialog = EditSettingDialog(ORGANIZATION_NAME, ALLICATION_NAME, DATABASE_FILE_NAME, TEMP_DB_NAME, parent=parent)
        result = dialog.exec_()
        new_settings = dialog._getSettings()
        # print dialog.spinBox_label_query_chunk_size.value(), dialog.spinBox_model_max_items.value(), result
        if result == QDialog.Accepted:
            settings = QSettings(QSettings.IniFormat, QSettings.UserScope,
                                 dialog.ORGANIZATION_NAME, dialog.ALLICATION_NAME)
            settings.setValue('Query_Chunk_Size', dialog.spinBox_label_query_chunk_size.value())
            settings.setValue('Max_Items_in_List', dialog.spinBox_model_max_items.value())
            settings.setValue('Query_limit', dialog.spinBox_query_limit.value())

            settings.setValue('Database_File_Name',         dialog.lineEdit_Database_File_Name.text())
            settings.setValue('Temp_Database_File_Name',    dialog.lineEdit_Temp_Database_File_Name.text())
            settings.sync()
        return (new_settings, result == QDialog.Accepted)

if __name__ == '__main__':
    import sys

    app = QtWidgets.QApplication(sys.argv)
    Dialog = QtWidgets.QDialog()
    ui = EditSettingDialog()
    ui.setupUi(Dialog)
    Dialog.show()
    sys.exit(app.exec_())

File: src/ui_change_advanced_setting_dialog.cpp
#include "ui_change_advanced_setting_dialog.h"
#include <QThread>

Dialog_Advanced_Setting::Dialog_Advanced_Setting(QWidget *parent):QDialog(parent)
{
    setupUi(this);

    QSettings settings(QSettings::IniFormat,QSettings::UserScope,
                       ORGANIZATION_NAME,ALLICATION_NAME);

    this->spinBox_label_query_chunk_size->setValue(
                settings.value("Query_Chunk_Size",10000).toInt());

    this->spinBox_model_max_items->setValue(
                settings.value("Max_Items_in_List",3000).toInt()
                );
    this->spinBox_query_limit->setValue(
                settings.value("Query_limit", 100).toInt()
                );

    this->spinBox_mount_state_update_interval->setValue(
                settings.value("Mount_State_Update_Interval", 3000).toInt()
                );
    this->spinBox_rowid_update_interval->setValue(
                settings.value("Rowid_Update_Interval",3000).toInt()
                );
    this->spinBox_db_update_interval->setValue(
                settings.value("Database_Update_Interval", 1000).toInt()
                );

    this->spinBox_lazy_query_interval->setValue(
                settings.value("Start_Querying_after_Typing_Finished", 50).toInt()
                );

    this->lineEdit_Database_File_Name->setText(DATABASE_FILE_NAME);
    this->lineEdit_Temp_Database_File_Name->setText(TEMP_DB_NAME);
    connect(this->lineEdit_Database_File_Name, SIGNAL(textEdited(QString)),
            this, SLOT(_on_lineEdit_Database_File_Name_edited(QString)));

    connect(this->lineEdit_Temp_Database_File_Name, SIGNAL(textEdited(QString)),
            this, SLOT(_on_Temp_lineEdit_Database_File_Name_edited(QString)));



    connect(this->toolButton_Database_File_Name, SIGNAL(released()),
            this, SLOT(_on_change_Database_File()));

    connect(this->toolButton_Temp_Database_File_Name, SIGNAL(released()),
            this, SLOT(_on_change_Temp_Database_File()));

    this->lineEdit_datetime_format->setText(DATETIME_FORMAT);
    this->checkBox_skip_diff_dev->setChecked(
                settings.value("Database/Skip_Different_Device", True).toBool());
    this->spinBox_restor_sort_interval->setValue(
                settings.value("Restor_Sort_after_New_Row_Inserted", 100).toInt()
                );
    QString unit = settings.value("Size_Unit", "KB").toString();
    QStringList unit_list = {"Auto","B", "KB", "MB", "GB", "TB", "PB"};
    int idx;
    if (!unit_list.contains(unit))
        idx =0 ;
    else
        idx = unit_list.indexOf(unit);

    this->comboBox_size_unit->setCurrentIndex(idx);

    if ( settings.value("Search/Instant_Search", True).toBool())
        this->radioButton_instant_search->setChecked(true);
    else
        this->radioButton_enter_to_search->setChecked(true);

    comboBox_threads_for_querying->clear();
    comboBox_threads_for_querying->addItem(QApplication::translate("dialog","All Cores"));
    for(int i =0; i<QThread::idealThreadCount(); i++)
    {
        comboBox_threads_for_querying->addItem(QString::number(i+1));
    }
    int query_thread = settings.value("Search/Threads", 0).toInt();
    if (query_thread< comboBox_threads_for_querying->count())
        comboBox_threads_for_querying->setCurrentIndex(query_thread);

    checkBox_compress_db_file->setChecked(COMPRESS_DB_FILE);

    lineEdit_ThemeName->setText(settings.value("General/Theme_Name", QIcon::themeName()).toString());
#if QT_VERSION >= QT_VERSION_CHECK(5, 12, 0)
    lineEdit_Fallback_ThemeName->setText(settings.value("General/Fallback_Theme_Name", QIcon::fallbackThemeName()).toString());
#else
    // lineEdit_Fallback_ThemeName->setText(settings.value("General/Fallback_Theme_Name", QIcon::themeName()).toString());
    lineEdit_Fallback_ThemeName->setDisabled(true);
#endif

    lineEdit_excluded_mount_path_regex->setText(settings.value("Database/Excluded_Mount_Path_Regex", DEFAULT_EXCLUDED_MOUNT_PATH_RE_STRING).toString());
}



void Dialog_Advanced_Setting::_on_lineEdit_Database_File_Name_edited(QString filename)
{
    QFileInfo path(filename);
    if (QFileInfo::exists(path.path()) && QFileInfo(path.path()).isDir())
        this->lineEdit_Database_File_Name->setStyleSheet("");
    else
        this->lineEdit_Database_File_Name->setStyleSheet("QLineEdit { color: red;  }");
}

void Dialog_Advanced_Setting::_on_Temp_lineEdit_Database_File_Name_edited(QString filename)
{
    QFileInfo path(filename);
    if (QFileInfo::exists(path.path()) && QFileInfo(path.path()).isDir())
        this->lineEdit_Temp_Database_File_Name->setStyleSheet("");
    else
        this->lineEdit_Temp_Database_File_Name->setStyleSheet("QLineEdit { color: red;  }");
}

void Dialog_Advanced_Setting::_on_change_Database_File()
{
   QString filename = QFileDialog::getSaveFileName(this,
                                                      QCoreApplication::translate("dialog","Select Directory of Database File"),
                            this->lineEdit_Database_File_Name->text()
        );
   if (filename!="")
       this->lineEdit_Database_File_Name->setText(filename);

}

void Dialog_Advanced_Setting::_on_change_Temp_Database_File()
{
    QString filename = QFileDialog::getSaveFileName(this,
                                                       QCoreApplication::translate("dialog","Select Directory of Database File"),
                             this->lineEdit_Temp_Database_File_Name->text()
         );
    if (filename!="")
        this->lineEdit_Temp_Database_File_Name->setText(filename);
}
bool Dialog_Advanced_Setting::getSettings(QWidget * parent)
{
    Dialog_Advanced_Setting dialog(parent);
    int result = dialog.exec();

    if (result == QDialog::Accepted)
    {
        QSettings settings(QSettings::IniFormat,QSettings::UserScope,
                           ORGANIZATION_NAME,ALLICATION_NAME);

        QUERY_CHUNK_SIZE = dialog.spinBox_label_query_chunk_size->value();
        settings.setValue("Query_Chunk_Size", dialog.spinBox_label_query_chunk_size->value());

        MODEL_MAX_ITEMS = dialog.spinBox_model_max_items->value();
        settings.setValue("Max_Items_in_List", dialog.spinBox_model_max_items->value());

        QUERY_LIMIT = dialog.spinBox_query_limit->value();
        settings.setValue("Query_limit", dialog.spinBox_query_limit->value());


        settings.setValue("Database_File_Name", dialog.lineEdit_Database_File_Name->text());
        settings.setValue("Temp_Database_File_Name", dialog.lineEdit_Temp_Database_File_Name->text());

        MOUNT_STATE_UPDATE_INTERVAL = dialog.spinBox_mount_state_update_interval->value();
        settings.setValue("Mount_State_Update_Interval", dialog.spinBox_mount_state_update_interval->value());
        ROWID_UPDATE_INTERVAL = dialog.spinBox_rowid_update_interval->value();
        settings.setValue("Rowid_Update_Interval", dialog.spinBox_rowid_update_interval->value());
        DB_UPDATE_INTERVAL = dialog.spinBox_db_update_interval->value();
        settings.setValue("Database_Update_Interval", dialog.spinBox_db_update_interval->value());


        DATETIME_FORMAT = dialog.lineEdit_datetime_format->text();
        settings.setValue("Search/Date_Format",
                          dialog.lineEdit_datetime_format->text());

        SKIP_DIFF_DEV = dialog.checkBox_skip_diff_dev->isChecked();
        settings.setValue("Database/Skip_Different_Device",
                          dialog.checkBox_skip_diff_dev->isChecked());


        QStringList unit_list = {"Auto","B", "KB", "MB", "GB", "TB", "PB"};
        QString unit =  unit_list[dialog.comboBox_size_unit->currentIndex()];
        settings.setValue("Size_Unit",unit);
        SIZE_UNIT = unit;

        QUERY_THREAD_NUMBER = dialog.comboBox_threads_for_querying->currentIndex();
        settings.setValue("Search/Threads", QUERY_THREAD_NUMBER);

        INSTANT_SEARCH = dialog.radioButton_instant_search->isChecked();
        settings.setValue("Search/Instant_Search", dialog.radioButton_instant_search->isChecked());

        settings.setValue("Start_Querying_after_Typing_Finished",dialog.spinBox_lazy_query_interval->value());
        settings.setValue("Restor_Sort_after_New_Row_Inserted", dialog.spinBox_restor_sort_interval->value());

        COMPRESS_DB_FILE = dialog.checkBox_compress_db_file->isChecked();
        settings.setValue("Compress_DB_File", COMPRESS_DB_FILE);

        QString themename = dialog.lineEdit_ThemeName->text();
        QString fallback_themename = dialog.lineEdit_Fallback_ThemeName->text();

        settings.setValue("General/Theme_Name", themename);
        settings.setValue("General/Fallback_Theme_Name", fallback_themename);

        QIcon::setThemeName(themename);
    #if QT_VERSION >= QT_VERSION_CHECK(5, 12, 0)
        QIcon::setFallbackThemeName(fallback_themename);
    #endif

        settings.setValue("Database/Excluded_Mount_Path_Regex", dialog.lineEdit_excluded_mount_path_regex->text());
        EXCLUDED_MOUNT_PATH_RE.setPattern(dialog.lineEdit_excluded_mount_path_regex->text());

        settings.sync();
        return true;
    }
    return false;
}
